File: plc4go/internal/plc4go/ads/readwrite/model/DataItem.go
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
package model

import (
	"errors"
	"github.com/apache/plc4x/plc4go/internal/plc4go/spi/utils"
	"github.com/apache/plc4x/plc4go/internal/plc4go/spi/values"
	api "github.com/apache/plc4x/plc4go/pkg/plc4go/values"
)

// Code generated by build-utils. DO NOT EDIT.

func DataItemParse(io *utils.ReadBuffer, dataFormatName string, stringLength int32) (api.PlcValue, error) {
	switch {
	case dataFormatName == "IEC61131_BOOL": // BOOL

		// Reserved Field (Just skip the bytes)
		if _, _err := io.ReadUint8(7); _err != nil {
			return nil, errors.New("Error parsing reserved field " + _err.Error())
		}

		// Simple Field (value)
		value, _valueErr := io.ReadBit()
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcBOOL(value), nil
	case dataFormatName == "IEC61131_BYTE": // BitString

		// Simple Field (value)
		value, _valueErr := io.ReadUint8(8)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcBitString(value), nil
	case dataFormatName == "IEC61131_WORD": // BitString

		// Simple Field (value)
		value, _valueErr := io.ReadUint16(16)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcBitString(value), nil
	case dataFormatName == "IEC61131_DWORD": // BitString

		// Simple Field (value)
		value, _valueErr := io.ReadUint32(32)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcBitString(value), nil
	case dataFormatName == "IEC61131_SINT": // SINT

		// Simple Field (value)
		value, _valueErr := io.ReadInt8(8)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcSINT(value), nil
	case dataFormatName == "IEC61131_USINT": // USINT

		// Simple Field (value)
		value, _valueErr := io.ReadUint8(8)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcUSINT(value), nil
	case dataFormatName == "IEC61131_INT": // INT

		// Simple Field (value)
		value, _valueErr := io.ReadInt16(16)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcINT(value), nil
	case dataFormatName == "IEC61131_UINT": // UINT

		// Simple Field (value)
		value, _valueErr := io.ReadUint16(16)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcUINT(value), nil
	case dataFormatName == "IEC61131_DINT": // DINT

		// Simple Field (value)
		value, _valueErr := io.ReadInt32(32)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcDINT(value), nil
	case dataFormatName == "IEC61131_UDINT": // UDINT

		// Simple Field (value)
		value, _valueErr := io.ReadUint32(32)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcUDINT(value), nil
	case dataFormatName == "IEC61131_LINT": // LINT

		// Simple Field (value)
		value, _valueErr := io.ReadInt64(64)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcLINT(value), nil
	case dataFormatName == "IEC61131_ULINT": // ULINT

		// Simple Field (value)
		value, _valueErr := io.ReadUint64(64)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcULINT(value), nil
	case dataFormatName == "IEC61131_REAL": // REAL

		// Simple Field (value)
		value, _valueErr := io.ReadFloat32(true, 8, 23)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcREAL(value), nil
	case dataFormatName == "IEC61131_LREAL": // LREAL

		// Simple Field (value)
		value, _valueErr := io.ReadFloat64(true, 11, 52)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcLREAL(value), nil
	case dataFormatName == "IEC61131_CHAR": // STRING
	case dataFormatName == "IEC61131_WCHAR": // STRING
	case dataFormatName == "IEC61131_STRING": // STRING

		// Manual Field (value)
		value, _valueErr := StaticHelperParseAmsString(io, stringLength, "UTF-8")
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcSTRING(value), nil
	case dataFormatName == "IEC61131_WSTRING": // STRING

		// Manual Field (value)
		value, _valueErr := StaticHelperParseAmsString(io, stringLength, "UTF-16")
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcSTRING(value), nil
	case dataFormatName == "IEC61131_TIME": // TIME

		// Simple Field (value)
		value, _valueErr := io.ReadUint32(32)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcTIME(value), nil
	case dataFormatName == "IEC61131_LTIME": // LTIME

		// Simple Field (value)
		value, _valueErr := io.ReadInt64(64)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcLTIME(value), nil
	case dataFormatName == "IEC61131_DATE": // DATE

		// Simple Field (value)
		value, _valueErr := io.ReadUint32(32)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcDATE(value), nil
	case dataFormatName == "IEC61131_TIME_OF_DAY": // TIME_OF_DAY

		// Simple Field (value)
		value, _valueErr := io.ReadUint32(32)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcTIME_OF_DAY(value), nil
	case dataFormatName == "IEC61131_DATE_AND_TIME": // DATE_AND_TIME

		// Simple Field (value)
		value, _valueErr := io.ReadUint32(32)
		if _valueErr != nil {
			return nil, errors.New("Error parsing 'value' field " + _valueErr.Error())
		}
		return values.NewPlcDATE_AND_TIME(value), nil
	}
	return nil, errors.New("unsupported type")
}

func DataItemSerialize(io *utils.WriteBuffer, value api.PlcValue, dataFormatName string, stringLength int32) error {
	switch {
	case dataFormatName == "IEC61131_BOOL": // BOOL

		// Reserved Field (Just skip the bytes)
		if _err := io.WriteUint8(7, uint8(0x00)); _err != nil {
			return errors.New("Error serializing reserved field " + _err.Error())
		}

		// Simple Field (value)
		if _err := io.WriteBit(value.GetBool()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_BYTE": // BitString

		// Simple Field (value)
		if _err := io.WriteUint8(8, value.GetUint8()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_WORD": // BitString

		// Simple Field (value)
		if _err := io.WriteUint16(16, value.GetUint16()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_DWORD": // BitString

		// Simple Field (value)
		if _err := io.WriteUint32(32, value.GetUint32()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_SINT": // SINT

		// Simple Field (value)
		if _err := io.WriteInt8(8, value.GetInt8()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_USINT": // USINT

		// Simple Field (value)
		if _err := io.WriteUint8(8, value.GetUint8()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_INT": // INT

		// Simple Field (value)
		if _err := io.WriteInt16(16, value.GetInt16()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_UINT": // UINT

		// Simple Field (value)
		if _err := io.WriteUint16(16, value.GetUint16()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_DINT": // DINT

		// Simple Field (value)
		if _err := io.WriteInt32(32, value.GetInt32()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_UDINT": // UDINT

		// Simple Field (value)
		if _err := io.WriteUint32(32, value.GetUint32()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_LINT": // LINT

		// Simple Field (value)
		if _err := io.WriteInt64(64, value.GetInt64()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_ULINT": // ULINT

		// Simple Field (value)
		if _err := io.WriteUint64(64, value.GetUint64()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_REAL": // REAL

		// Simple Field (value)
		if _err := io.WriteFloat32(32, value.GetFloat32()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_LREAL": // LREAL

		// Simple Field (value)
		if _err := io.WriteFloat64(64, value.GetFloat64()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_CHAR": // STRING
	case dataFormatName == "IEC61131_WCHAR": // STRING
	case dataFormatName == "IEC61131_STRING": // STRING

		// Manual Field (value)
		_valueErr := StaticHelperSerializeAmsString(io, value, stringLength, "UTF-8")
		if _valueErr != nil {
			return errors.New("Error serializing 'value' field " + _valueErr.Error())
		}
	case dataFormatName == "IEC61131_WSTRING": // STRING

		// Manual Field (value)
		_valueErr := StaticHelperSerializeAmsString(io, value, stringLength, "UTF-16")
		if _valueErr != nil {
			return errors.New("Error serializing 'value' field " + _valueErr.Error())
		}
	case dataFormatName == "IEC61131_TIME": // TIME

		// Simple Field (value)
		if _err := io.WriteUint32(32, value.GetUint32()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_LTIME": // LTIME

		// Simple Field (value)
		if _err := io.WriteInt64(64, value.GetInt64()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_DATE": // DATE

		// Simple Field (value)
		if _err := io.WriteUint32(32, value.GetUint32()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_TIME_OF_DAY": // TIME_OF_DAY

		// Simple Field (value)
		if _err := io.WriteUint32(32, value.GetUint32()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	case dataFormatName == "IEC61131_DATE_AND_TIME": // DATE_AND_TIME

		// Simple Field (value)
		if _err := io.WriteUint32(32, value.GetUint32()); _err != nil {
			return errors.New("Error serializing 'value' field " + _err.Error())
		}
	default:

		return errors.New("unsupported type")
	}
	return nil
}

File: sandbox/plc4py/plc4py/protocols/umas/readwrite/DataItem.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

from plc4py.api.value.PlcValue import PlcValue
from plc4py.protocols.umas import StaticHelper
from plc4py.protocols.umas.readwrite.UmasDataType import UmasDataType
from plc4py.spi.generation.ReadBuffer import ReadBuffer
from plc4py.spi.generation.WriteBuffer import WriteBuffer
from plc4py.spi.values.PlcValues import PlcBOOL
from plc4py.spi.values.PlcValues import PlcBYTE
from plc4py.spi.values.PlcValues import PlcDINT
from plc4py.spi.values.PlcValues import PlcDWORD
from plc4py.spi.values.PlcValues import PlcINT
from plc4py.spi.values.PlcValues import PlcList
from plc4py.spi.values.PlcValues import PlcREAL
from plc4py.spi.values.PlcValues import PlcSTRING
from plc4py.spi.values.PlcValues import PlcUDINT
from plc4py.spi.values.PlcValues import PlcUINT
from plc4py.spi.values.PlcValues import PlcULINT
from plc4py.spi.values.PlcValues import PlcWORD
from plc4py.utils.GenericTypes import ByteOrder
from typing import List
import logging
import math


class DataItem:
    @staticmethod
    def static_parse(
        read_buffer: ReadBuffer, data_type: UmasDataType, number_of_values: int
    ):
        if data_type == UmasDataType.BOOL and number_of_values == int(1):  # BOOL

            # Reserved Field (Compartmentalized so the "reserved" variable can't leak)
            reserved: int = read_buffer.read_unsigned_short(7, logical_name="")
            if reserved != int(0x0000):
                logging.warning(
                    "Expected constant value "
                    + str(0x0000)
                    + " but got "
                    + str(reserved)
                    + " for reserved field."
                )

            # Simple Field (value)
            value: bool = read_buffer.read_bit("")

            return PlcBOOL(value)
        if data_type == UmasDataType.BOOL:  # List
            # Array field (value)
            # Count array
            item_count: int = int(number_of_values)
            value: List[PlcValue] = []
            for cur_item in range(item_count):
                value.append(PlcBOOL(bool(read_buffer.read_bit(""))))

            return PlcList(value)
        if data_type == UmasDataType.BYTE and number_of_values == int(1):  # BYTE

            # Simple Field (value)
            value: int = read_buffer.read_unsigned_short(8, logical_name="")

            return PlcBYTE(value)
        if data_type == UmasDataType.BYTE:  # List
            # Array field (value)
            # Count array
            item_count: int = int(number_of_values * int(8))
            value: List[PlcValue] = []
            for cur_item in range(item_count):
                value.append(PlcBOOL(bool(read_buffer.read_bit(""))))

            return PlcList(value)
        if data_type == UmasDataType.WORD:  # WORD

            # Simple Field (value)
            value: int = read_buffer.read_unsigned_int(16, logical_name="")

            return PlcWORD(value)
        if data_type == UmasDataType.DWORD:  # DWORD

            # Simple Field (value)
            value: int = read_buffer.read_unsigned_long(32, logical_name="")

            return PlcDWORD(value)
        if data_type == UmasDataType.INT and number_of_values == int(1):  # INT

            # Simple Field (value)
            value: int = read_buffer.read_short(16, logical_name="")

            return PlcINT(value)
        if data_type == UmasDataType.INT:  # List
            # Array field (value)
            # Count array
            item_count: int = int(number_of_values)
            value: List[PlcValue] = []
            for cur_item in range(item_count):
                value.append(PlcINT(int(read_buffer.read_short(16, logical_name=""))))

            return PlcList(value)
        if data_type == UmasDataType.DINT and number_of_values == int(1):  # DINT

            # Simple Field (value)
            value: int = read_buffer.read_int(32, logical_name="")

            return PlcDINT(value)
        if data_type == UmasDataType.DINT:  # List
            # Array field (value)
            # Count array
            item_count: int = int(number_of_values)
            value: List[PlcValue] = []
            for cur_item in range(item_count):
                value.append(PlcDINT(int(read_buffer.read_int(32, logical_name=""))))

            return PlcList(value)
        if data_type == UmasDataType.UINT and number_of_values == int(1):  # UINT

            # Simple Field (value)
            value: int = read_buffer.read_unsigned_int(16, logical_name="")

            return PlcUINT(value)
        if data_type == UmasDataType.UINT:  # List
            # Array field (value)
            # Count array
            item_count: int = int(number_of_values)
            value: List[PlcValue] = []
            for cur_item in range(item_count):
                value.append(
                    PlcUDINT(int(read_buffer.read_unsigned_int(16, logical_name="")))
                )

            return PlcList(value)
        if data_type == UmasDataType.UDINT and number_of_values == int(1):  # UDINT

            # Simple Field (value)
            value: int = read_buffer.read_unsigned_long(32, logical_name="")

            return PlcUDINT(value)
        if data_type == UmasDataType.UDINT:  # List
            # Array field (value)
            # Count array
            item_count: int = int(number_of_values)
            value: List[PlcValue] = []
            for cur_item in range(item_count):
                value.append(
                    PlcULINT(int(read_buffer.read_unsigned_long(32, logical_name="")))
                )

            return PlcList(value)
        if data_type == UmasDataType.REAL and number_of_values == int(1):  # REAL

            # Simple Field (value)
            value: float = read_buffer.read_float(32, logical_name="")

            return PlcREAL(value)
        if data_type == UmasDataType.REAL:  # List
            # Array field (value)
            # Count array
            item_count: int = int(number_of_values)
            value: List[PlcValue] = []
            for cur_item in range(item_count):
                value.append(
                    PlcREAL(float(read_buffer.read_float(32, logical_name="")))
                )

            return PlcList(value)
        if data_type == UmasDataType.STRING and number_of_values == int(1):  # STRING
            # Manual Field (value)
            value: str = (str)(
                StaticHelper.parse_terminated_string_bytes(
                    read_buffer, number_of_values
                )
            )

            return PlcSTRING(value)
        if data_type == UmasDataType.STRING:  # List
            # Array field (value)
            # Count array
            item_count: int = int(number_of_values)
            value: List[PlcValue] = []
            for cur_item in range(item_count):
                value.append(
                    PlcREAL(float(read_buffer.read_float(32, logical_name="")))
                )

            return PlcList(value)
        return None

    @staticmethod
    def static_serialize(
        write_buffer: WriteBuffer,
        _value: PlcValue,
        data_type: UmasDataType,
        number_of_values: int,
        byte_order: ByteOrder,
    ) -> None:
        if data_type == UmasDataType.BOOL and number_of_values == int(1):  # BOOL

            # Reserved Field
            write_buffer.write_byte(int(0x0000), 7, "int0x0000")
            # Simple Field (value)
            value: bool = _value.get_bool()
            write_buffer.write_bit((value), "value")
        if data_type == UmasDataType.BOOL:  # List

            values: PlcList = _value

            for val in values.get_list():
                value: bool = val.get_bool()
                write_buffer.write_bit((value), "value")

        if data_type == UmasDataType.BYTE and number_of_values == int(1):  # BYTE

            # Simple Field (value)
            value: int = _value.get_int()
            write_buffer.write_byte((value), 8, "value")
        if data_type == UmasDataType.BYTE:  # List

            values: PlcList = _value

            for val in values.get_list():
                value: bool = val.get_bool()
                write_buffer.write_bit((value), "value")

        if data_type == UmasDataType.WORD:  # WORD

            # Simple Field (value)
            value: int = _value.get_int()
            write_buffer.write_unsigned_short((value), 16, "value")
        if data_type == UmasDataType.DWORD:  # DWORD

            # Simple Field (value)
            value: int = _value.get_int()
            write_buffer.write_unsigned_int((value), 32, "value")
        if data_type == UmasDataType.INT and number_of_values == int(1):  # INT

            # Simple Field (value)
            value: int = _value.get_int()
            write_buffer.write_short((value), 16, "value")
        if data_type == UmasDataType.INT:  # List

            values: PlcList = _value

            for val in values.get_list():
                value: int = val.get_int()
                write_buffer.write_short((value), 16, "value")

        if data_type == UmasDataType.DINT and number_of_values == int(1):  # DINT

            # Simple Field (value)
            value: int = _value.get_int()
            write_buffer.write_int((value), 32, "value")
        if data_type == UmasDataType.DINT:  # List

            values: PlcList = _value

            for val in values.get_list():
                value: int = val.get_int()
                write_buffer.write_int((value), 32, "value")

        if data_type == UmasDataType.UINT and number_of_values == int(1):  # UINT

            # Simple Field (value)
            value: int = _value.get_int()
            write_buffer.write_unsigned_short((value), 16, "value")
        if data_type == UmasDataType.UINT:  # List

            values: PlcList = _value

            for val in values.get_list():
                value: int = val.get_int()
                write_buffer.write_unsigned_short((value), 16, "value")

        if data_type == UmasDataType.UDINT and number_of_values == int(1):  # UDINT

            # Simple Field (value)
            value: int = _value.get_int()
            write_buffer.write_unsigned_int((value), 32, "value")
        if data_type == UmasDataType.UDINT:  # List

            values: PlcList = _value

            for val in values.get_list():
                value: int = val.get_int()
                write_buffer.write_unsigned_int((value), 32, "value")

        if data_type == UmasDataType.REAL and number_of_values == int(1):  # REAL

            # Simple Field (value)
            value: float = _value.get_float()
            write_buffer.write_float((value), 32, "value")
        if data_type == UmasDataType.REAL:  # List

            values: PlcList = _value

            for val in values.get_list():
                value: float = val.get_float()
                write_buffer.write_float((value), 32, "value")

        if data_type == UmasDataType.STRING and number_of_values == int(1):  # STRING

            # Manual Field (value)
            serialize_terminated_string(write_buffer, self.value, self.number_of_values)
        if data_type == UmasDataType.STRING:  # List

            values: PlcList = _value

            for val in values.get_list():
                value: float = val.get_float()
                write_buffer.write_float((value), 32, "value")

    @staticmethod
    def get_length_in_bytes(
        _value: PlcValue, data_type: UmasDataType, number_of_values: int
    ) -> int:
        return int(
            math.ceil(
                float(DataItem.get_length_in_bits(_value, data_type, number_of_values))
                / 8.0
            )
        )

    @staticmethod
    def get_length_in_bits(
        _value: PlcValue, data_type: UmasDataType, number_of_values: int
    ) -> int:
        size_in_bits: int = 0
        if data_type == UmasDataType.BOOL and number_of_values == int(1):  # BOOL
            # Reserved Field
            size_in_bits += 7
            # Simple Field (value)
            size_in_bits += 1
        if data_type == UmasDataType.BOOL:  # List
            values: PlcList = _value
            size_in_bits += len(values.get_list()) * 1
        if data_type == UmasDataType.BYTE and number_of_values == int(1):  # BYTE
            # Simple Field (value)
            size_in_bits += 8
        if data_type == UmasDataType.BYTE:  # List
            values: PlcList = _value
            size_in_bits += len(values.get_list()) * 1
        if data_type == UmasDataType.WORD:  # WORD
            # Simple Field (value)
            size_in_bits += 16
        if data_type == UmasDataType.DWORD:  # DWORD
            # Simple Field (value)
            size_in_bits += 32
        if data_type == UmasDataType.INT and number_of_values == int(1):  # INT
            # Simple Field (value)
            size_in_bits += 16
        if data_type == UmasDataType.INT:  # List
            values: PlcList = _value
            size_in_bits += len(values.get_list()) * 16
        if data_type == UmasDataType.DINT and number_of_values == int(1):  # DINT
            # Simple Field (value)
            size_in_bits += 32
        if data_type == UmasDataType.DINT:  # List
            values: PlcList = _value
            size_in_bits += len(values.get_list()) * 32
        if data_type == UmasDataType.UINT and number_of_values == int(1):  # UINT
            # Simple Field (value)
            size_in_bits += 16
        if data_type == UmasDataType.UINT:  # List
            values: PlcList = _value
            size_in_bits += len(values.get_list()) * 16
        if data_type == UmasDataType.UDINT and number_of_values == int(1):  # UDINT
            # Simple Field (value)
            size_in_bits += 32
        if data_type == UmasDataType.UDINT:  # List
            values: PlcList = _value
            size_in_bits += len(values.get_list()) * 32
        if data_type == UmasDataType.REAL and number_of_values == int(1):  # REAL
            # Simple Field (value)
            size_in_bits += 32
        if data_type == UmasDataType.REAL:  # List
            values: PlcList = _value
            size_in_bits += len(values.get_list()) * 32
        if data_type == UmasDataType.STRING and number_of_values == int(1):  # STRING
            # Manual Field (value)
            size_in_bits += self.number_of_values * int(8)
        if data_type == UmasDataType.STRING:  # List
            values: PlcList = _value
            size_in_bits += len(values.get_list()) * 32
        return size_in_bits
